system-interface/src/instruction.rs

```diff
@@ -249,6 +256,55 @@ pub enum SystemInstruction {
     /// # Account references
     ///   0. `[WRITE]` Nonce account
     UpgradeNonceAccount,
+
+    /// Create a new account without enforcing the invariant that the account's
+    /// current lamports must be 0.
+    ///
+    /// This constructor is identical to [`create_account`] with the exception that it
+    /// **does not** check that the destination account (`to_pubkey`) has a zero
+    /// lamport balance prior to creation. This enables patterns where you first transfer
+    /// lamports to prefund an account, then use `create_account_allow_prefund` as a single
+    /// CPI to transfer additional lamports, allocate space, and assign ownership.
+    ///
+    /// Use [`create_account`] for typical account creation.
+    /// Use [`create_account_allow_prefund`] when the target account has already been
+    /// prefunded and you want to complete the creation process with a single CPI.
+    ///
+    /// **Safety considerations**
+    /// As with `allocate` and `assign` when invoked manually, this instruction can brick
+    /// a wallet if used incorrectly; do not pass in a wallet system account as the new
+    /// account. This instruction does not prevent the new account from having more
+    /// lamports than required for rent exemption, and all lamports will become locked.
+    ///
+    /// # Account references
+    /// If `lamports > 0` (meaning lamports are being transferred):
+    ///   0. `[WRITE, SIGNER]` New account
+    ///   1. `[WRITE, SIGNER]` Funding account
+    ///
+    /// If `lamports == 0` (no lamports to be transferred), you may omit funding account:
+    ///   0. `[WRITE, SIGNER]` New account
+    CreateAccountAllowPrefund {
+        /// Number of lamports to transfer to the new account
+        lamports: u64,
+
+        /// Number of bytes of memory to allocate
+        space: u64,
+
+        /// Address of program that will own the new account
+        owner: Address,
+    },
+}
+
+/// A funding argument where any amount of `lamports` requires a `from`
+/// address. Currently only for use with `create_account_allow_prefund`.
+#[cfg_attr(
+    feature = "serde",
+    derive(serde_derive::Deserialize, serde_derive::Serialize)
+)]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Funding {
+    None,
+    Transfer { from: Address, lamports: u64 },
```

Comment:
To be honest, I'm not sure if this is a great idea. It's currently passed in by value, which forces a copy of the pubkey and lamports.

Even if it's passed by reference, users are forced to create an intermediate struct just to create the instruction.

Finally, the resulting code is more verbose than just taking in `lamports` and `Option<&Address>` always.

Is that worth the tradeoff to make it impossible for someone to specify an amount of lamports without also specifying a funder? Maybe! But I'm not convinced.

Reply:
Cc @grod220 - how strongly do you feel on the `Funding` type protection here? 

I love such things, but we were already checking the `Option<Address>` in the original anyway:

So for better UX than failing downstream, I can just error here, no extra work, no extra type.